pkg/cache/v3/linear.go

```diff
@@ -29,37 +30,86 @@ import (
 // cachedResource is used to track resources added by the user in the cache.
 // It contains the resource itself and its associated version (currently in two different modes).
 type cachedResource struct {
-	types.Resource
+	name string
+
+	resource types.Resource
+	ttl      *time.Duration
 
 	// cacheVersion is the version of the cache at the time of last update, used in sotw.
 	cacheVersion string
+
 	// stableVersion is the version of the resource itself (a hash of its content after deterministic marshaling).
 	// It is lazy initialized and should be accessed through getStableVersion.
 	stableVersion string
+
+	// marshaledResource contains the marshaled version of the resource.
+	// It is lazy initialized and should be accessed through getMarshaledResource
+	marshaledResource []byte
+
+	// mu is the mutex used to lazy compute the marshaled resource and stable version.
+	mu sync.Mutex
```

Comment:
This could be simplified quite a bit by using `sync.Once` with a custom struct that ensures we compute the stable version and the serialized proto once:

```
[...]
	// marshalResource computes the resource bytes and version...
	marshalResource func() (ser, error)
}
```

In the initialization:

```
[...]
		marshalResource: sync.OnceValues(func() (s ser, err error) {
			s.res, err = MarshalResource(res)
			h := sha256.New()
			h.Write(c.res)
			s.v = hex.EncodeToString(h.Sum(nil))
			return c, err
		}),
```

And then you can freely call marshalResource anytime you need without worrying about locking.